### PR TITLE
Adding dynamic backends for the IEx H command.

### PR DESCRIPTION
**Pull request by bbense**
This pull request does two things:

First it adds a `:doc_helper` keyword list to the options for
`IEx.Config`. This contains two values to assist in allowing the
iex h command to have a pluggable backend.

The second is that it replaces the workings of the h command to
use a list of modules that implement a documention function. The
purpose of this is to allow the h command to potentially have access
to Erlang documentation. See:

#3589

There are still some outstanding issues:

1. The documentation interface needs spec and type settings. Now
that Behaviour is deprecated, I'm not sure what the correct thing
to do is wrt specific callbacks.

2. Should the elixir code include any optional documentation helpers?

After some false starts I believe this version fully duplicates the
behaviour of the current h command for Elixir documentation.

### CODE REVIEW

**Comment by josevalim** [comment]
> The documentation interface needs spec and type settings. Now
that Behaviour is deprecated, I'm not sure what the correct thing
to do is wrt specific callbacks.

I don't think we need to do anything here. Typespecs and callbacks work exactly the same between Erlang and Elixir modules. So everything should just work. :)

File: lib/iex/lib/iex/h_elixir.ex [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+  end 
+
+	#  match on all arities.
+	def find_doc(docs, module, function) do
+		doc_list = docs |> Enum.filter(fn(x) -> match_function(x, function) end) 
+		case doc_list do 
+			[] -> {:not_found, [{"#{inspect module}.#{function}", "No documentation for #{inspect module}.#{function} found\n"}]}
+			_  -> {:found, get_docstrings(doc_list)}
+		end 
+	end
```
Comment (ericmj) [inline lib/iex/lib/iex/h_elixir.ex]: Change tabs to spaces.

**Comment by ericmj** [comment]
Please force push instead of opening new PRs, looks like we lost some code comments from the old PR.

**Comment by dukex** [comment]
Hi @bbense can you fix the tabulation in lib/iex/lib/iex/h_elixir.ex file?

**Comment by bbense** [comment]
Ok, thought I had swatted that already, but I'll make another pass.

**Comment by eksperimental** [comment]
It will be really cool to emit s warning when compiling.. I see this happening a lot. And these are doc breaker errors

File: lib/iex/lib/iex/h_erlang_stub.ex [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  end
+
+  def is_erlang?(module) do
+    case is_elixir?(module) do
+      true -> false
+      _    -> true
+    end
+  end
+
+  def is_elixir?(module) do
```
Comment (mgwidmann) [inline lib/iex/lib/iex/h_erlang_stub.ex]: This function is an exact duplicate from `IEx.H_Elixir`. Perhaps `import IEx.H_Elixir, only: [is_elixir?: 1]` instead?

File: lib/iex/lib/iex/h_erlang_stub.ex [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+  def documentation(module, function,arity) do
+    case is_erlang?(module) do
+      true -> get_doc(module, function,arity)
+      _    -> {:unknown, [{inspect(module), ""}]}
+    end
+  end
+
+  def is_erlang?(module) do
+    case is_elixir?(module) do
```
Comment (mgwidmann) [inline lib/iex/lib/iex/h_erlang_stub.ex]: Is there any reason why this can't be just `not is_elixir?(module)`? 

Reminds me of seeing something like this from the OO world: `is_elixir? ? true : false` :laughing:

Comment (eksperimental) [inline lib/iex/lib/iex/h_erlang_stub.ex]: functions should only be named is_* in elixir if and only if they can be used in guards. and if they can, they should not end with "?".
for the rest of the functions that return a boolean, we use functions ending in "?".

for other "verbs", like "to have"  is ok to use both, verb and "?".. like Dict functions. has_key?/2

File: lib/iex/lib/iex/h_elixir.ex [hunk 4]
```diff
@@ -0,0 +1,143 @@
+defmodule IEx.H_Elixir do
```
Comment (lexmag) [inline lib/iex/lib/iex/h_elixir.ex]: This module name violates common naming.

File: lib/iex/lib/iex/h_erlang_stub.ex [hunk 5]
```diff
@@ -0,0 +1,57 @@
+defmodule IEx.H_Erlang_Stub do
```
Comment (lexmag) [inline lib/iex/lib/iex/h_erlang_stub.ex]: This module name violates common naming.

File: lib/iex/lib/iex.ex [hunk 6]
```diff
@@ -213,7 +213,7 @@ defmodule IEx do
   Configures IEx.
 
   The supported options are: `:colors`, `:inspect`,
-  `:default_prompt`, `:alive_prompt` and `:history_size`.
+  `:default_prompt`, `:alive_prompt` , `:doc_helpers` and `:history_size`.
```
Comment (eksperimental) [inline lib/iex/lib/iex.ex]: there's an extra space after \``:alive_prompt`\`

File: lib/iex/lib/iex.ex [hunk 7]
```diff
@@ -255,6 +255,28 @@ defmodule IEx do
...
+  Documentation Helpers allows iex to have a configurable set of modules
+  for looking up both Elixir and Erlang documentation. `:doc_helpers` is
+  a keyword list with these values.
+
+    * `:helpers` - A list of modules that implement the documentation
+                   functions. The default is `[IEx.H_Elixir, IEx.H_Erlang_Stub]`.
+                   Removing `IEx.H_Elixir` is not recommended.
+    * `:find`    - Can have two values `:first` or `:all` (:first is default)
+                   Determines whether all helpers are called or just
+                   the first helper that returns `:found`
```
Comment (eksperimental) [inline lib/iex/lib/iex.ex]: this was mentioned before i think, but it will not hurt to remind it again,
new lines should be idented with an extra 2 space, not to whatever spaces the previous line determines. if this works in the docs is sheer coincidence.

so the first block should read:
```elixir
    * `:helpers` - A list of modules that implement the documentation
      functions. The default is `[IEx.H_Elixir, IEx.H_Erlang_Stub]`.
      Removing `IEx.H_Elixir` is not recommended.
```

File: lib/iex/lib/iex.ex [hunk 8]
```diff
@@ -255,6 +255,28 @@ defmodule IEx do
...
+
+    An sample .iex.exs to use this would be
+
+    # .iex.exs
+    Code.ensure_loaded(Erlang.H_Helper)
+    IEx.configure(doc_helpers: [helpers: [IEx.H_Elixir, Erlang.H_Helper]])
+
+  See `IEx.H_Elixir` for the full description of the documentation functions.
```
Comment (eksperimental) [inline lib/iex/lib/iex.ex]: extra period in `functions.`

File: lib/iex/lib/iex/h_elixir.ex [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+  All of these functions return a tuple of status and doc_list.
+
+  * `{:found, doc_list}`     - Documentation was found
+
+  * `{:not_found, doc_list}` - Documentation was not found and
+                              this helper is sure it can't be found
+
+  * `{:unknown, doc_list}`   - Helper does not know how to find docs for
+                               arguments given.
+
```
Comment (eksperimental) [inline lib/iex/lib/iex/h_elixir.ex]: i think we shouldn't empty new lines in  between items in a list.

File: lib/iex/lib/iex/h_elixir.ex [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+      _ ->
+        false
+    end
+  end
+
+  defp match_function(docstring, function) do
+    {func, _arity} = elem(docstring, 0)
+    function == func
+  end
+
```
Comment (eksperimental) [inline lib/iex/lib/iex/h_elixir.ex]: since this funtion returns a boolean, it should be named "match_function?"

File: lib/iex/lib/iex/h_elixir.ex [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+    Atom.to_string(var)
+  end
+
+  defp find_default_doc(doc, function, min) do
+    case elem(doc, 0) do
+      {^function, max} when max > min ->
+        defaults = Enum.count elem(doc, 3), &match?({:\\, _, _}, &1)
+        min + defaults >= max
+      _ ->
+        false
```
Comment (eksperimental) [inline lib/iex/lib/iex/h_elixir.ex]: i don't get the relation btw the name of this function with what it returns (a boolean)

File: lib/iex/lib/iex/introspection.ex [hunk 12]
```diff
@@ -39,112 +30,183 @@ defmodule IEx.Introspection do
...
-        h_mod_fun(module, function) == :ok
-      end
-
-    unless result, do: nodocs(function)
+    helpers = IEx.Config.doc_helpers(:helpers)
+    result = Enum.find_value(modules,
+               fn(module) ->
+                  Enum.find_value(helpers,
+                        fn(mod) -> found_help(mod, module, function) end)
+               end)
```
Comment (eksperimental) [inline lib/iex/lib/iex/introspection.ex]: I find these indentation particularly hard to follow (it's just a personal choice), but i think the most prevalent way to identing in the Elixir project is just to add 2 spaces on a new level on the next line.
having an odd number of spaces also make it diffucult to line up using the tab space in editors that replace a tab with two spaces.

File: lib/iex/lib/iex/introspection.ex [hunk 13]
```diff
@@ -39,112 +30,183 @@ defmodule IEx.Introspection do
...
-    unless result, do: nodocs("#{function}/#{arity}")
+    helpers = IEx.Config.doc_helpers(:helpers)
+    result = Enum.find_value(modules,
+              fn(module) -> Enum.find_value(helpers,
+                             fn(mod) -> found_help(mod, module, function, arity) end)
+              end)
+
+    case result do
+      {:found, doc_list} -> display_doc_list(:found, doc_list)
+      _                  ->
```
Comment (eksperimental) [inline lib/iex/lib/iex/introspection.ex]: i would definitely remove those spaces.

File: lib/iex/lib/iex/introspection.ex (hunk 12 above)
Comment (mgwidmann) [inline lib/iex/lib/iex/introspection.ex]: :+1: I had similar trouble.

Comment (lexmag) [inline lib/iex/lib/iex/introspection.ex]: Strongly :+1: @eksperimental.

**Comment by lexmag** [comment]
@eksperimental thank you for reviewing.

File: lib/iex/lib/iex/introspection.ex (hunk 12 above)
Comment (whatyouhide) [inline lib/iex/lib/iex/introspection.ex]: There was a long discussion on the Elixir style guide on this. This is what this code would look like:

```elixir
result =
  Enum.find_value modules, fn(module) ->
    Enum.find_value(helpers, fn(mod) -> found_help(mod, module, function) end)
  end
```

Comment (lexmag) [inline lib/iex/lib/iex/introspection.ex]: @whatyouhide and probably worth to simplify it with the capture operator. :smile:

File: lib/iex/lib/iex.ex [hunk 14]
```diff
@@ -255,6 +255,28 @@ defmodule IEx do
   Number of expressions and their results to keep in the history.
   The value is an integer. When it is negative, the history is unlimited.
 
+  ## Documentation Helpers
+
+  Documentation Helpers allows iex to have a configurable set of modules
```
Comment (eksperimental) [inline lib/iex/lib/iex.ex]: It should read "Documentation Helpers allow IEx"
(change the tense of the "to allow" verb and capitalization of IEx)

File: lib/iex/lib/iex/h_elixir.ex [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+    case docs do
+      nil -> {:not_found, [{inspect(module), "No moduledocs found\n"}]}
+      _   -> {:found, [{inspect(module), elem(docs, 1)}]}
+    end
+  end
+
+  def get_doc(module, function) when is_atom(module) and is_atom(function) do
+    docs = Code.get_docs(module, :docs)
+    case docs do
+      nil -> {:not_found, [{"#{inspect module}.#{function}", "No documentation for #{inspect module}.#{function} found\n"}]}
```
Comment (eksperimental) [inline lib/iex/lib/iex/h_elixir.ex]: regarding all the :no_found messages, wouldn't it be more covenient to just add `\n` when printing it that add it every time you define the message? I mean, there are more changes that you will defining one of them, and missing it when you print it, and if the latter happens it will be easier to spot and fix.

File: lib/iex/lib/iex/introspection.ex [hunk 16]
```diff
@@ -39,112 +30,183 @@ defmodule IEx.Introspection do
...
 
     dont_display_result
   end
 
-  defp h_mod_fun_arity(mod, fun, arity) when is_atom(mod) do
-    if docs = Code.get_docs(mod, :docs) do
-      doc = find_doc(docs, fun, arity, 4)
-            || find_default_doc(docs, fun, arity)
+  @doc """
+  Map over helpers and return a list of header, doc tuples.
```
Comment (eksperimental) [inline lib/iex/lib/iex/introspection.ex]: Docs should use the verbs in Present continuos, not in impreative form,
so it should read
> Maps over helpers and returns a list of header, doc tuples.

File: lib/iex/lib/iex/introspection.ex [hunk 17]
```diff
@@ -39,112 +30,183 @@ defmodule IEx.Introspection do
...
   end
 
-  defp find_doc(docs, function, arity, pos) do
-    if doc = List.keyfind(docs, {function, arity}, 0) do
-      case elem(doc, pos) do
-        false -> nil
-        _ -> doc
-      end
+  @doc """
+  Return nil if mod.documentation returns :unknown
```
Comment (eksperimental) [inline lib/iex/lib/iex/introspection.ex]: verbs should in in present continuos, not in imperative form. it should read.
> Returns `nil` if mod.documentation returns :unknown

nil should have backticks, and I'm not sure if we should have backticks in `mod.documentation` (probably yes)

File: lib/iex/lib/iex/introspection.ex (hunk 12 above)
Comment (eksperimental) [inline lib/iex/lib/iex/introspection.ex]: thank you @whatyouhide that's a really neat solution. I didn't know about that one, but look clean and readable.

File: lib/iex/lib/iex/introspection.ex (hunk 16 above)
Comment (eksperimental) [inline lib/iex/lib/iex/introspection.ex]: also I would refer to the tuple this way:
> Maps over helpers and returns a list of `{header, doc}` tuples.

File: lib/iex/lib/iex/introspection.ex (hunk 12 above)
Comment (bbense) [inline lib/iex/lib/iex/introspection.ex]: I struggle how to format those kinds of nested iterations a lot. Are there
any examples
elsewhere in the code base ?

- Booker C. Bense

File: lib/iex/lib/iex.ex (hunk 7 above)
Comment (eksperimental) [inline lib/iex/lib/iex.ex]: we are not using backtick in documentation around atoms; since it is understood that they are values.
so the line above can read
> Can have two values :first or :all (:first is default)

what I'm not sure, is that if we should use backticks as the head of the item, such as in `:helpers` or `:find`, maybe @josevalim has a say on this?

**Comment by eksperimental** [comment]
@bbense I have submitted (quiet a few) comments, I hope you can be of help.

@lexmag no worries, I always find it amusing and a learning experience when you do that to my code ;-D

File: lib/iex/lib/iex/h_erlang_stub.ex (hunk 3 above)
Comment (bbense) [inline lib/iex/lib/iex/h_erlang_stub.ex]: Does this apply to private functions as well? Cause that should not be a public function. I'll fix it either way. Just curious.

File: lib/iex/lib/iex/h_elixir.ex (hunk 4 above)
Comment (bbense) [inline lib/iex/lib/iex/h_elixir.ex]: Ok, I'm thinking of changing them to 

IEx.DocHelp.Elixir, IEx.DocHelp.ErlangStub

or 

IEx.DocExilir , IEx.DocErlangStub

I'm thinking the first might be useful to abstract some common functionality between the doc helpers. This might address some of the other comments.

File: lib/iex/lib/iex.ex (hunk 6 above)
Comment (bbense) [inline lib/iex/lib/iex.ex]: fixed

File: lib/iex/lib/iex/h_elixir.ex (hunk 9 above)
Comment (bbense) [inline lib/iex/lib/iex/h_elixir.ex]: Spaces are used between the list items elsewhere, See Code.get_docs

File: lib/iex/lib/iex/h_elixir.ex (hunk 10 above)
Comment (bbense) [inline lib/iex/lib/iex/h_elixir.ex]: Jose wrote that code, I just copied it out of introspection.ex

File: lib/iex/lib/iex.ex (hunk 14 above)
Comment (bbense) [inline lib/iex/lib/iex.ex]: fixed

File: lib/iex/lib/iex/h_elixir.ex (hunk 9 above)
Comment (eksperimental) [inline lib/iex/lib/iex/h_elixir.ex]: yes, sorry about that one. It's used both ways. I just thought it could break the list, but it doesn't. so you are fine.

File: lib/iex/lib/iex/h_erlang_stub.ex (hunk 3 above)
Comment (eksperimental) [inline lib/iex/lib/iex/h_erlang_stub.ex]: probably yes. I don't see why have a different criteria for private functions. I think it can be source of confusion, specially when you invoke them, since when you see it somewhere else in the code, you don't know whether is private or public, and that make you believe that you can use it in guards

Comment (lexmag) [inline lib/iex/lib/iex/h_erlang_stub.ex]: It applies to every function regardless of visibility.

File: lib/iex/lib/iex/h_elixir.ex (hunk 4 above)
Comment (lexmag) [inline lib/iex/lib/iex/h_elixir.ex]: The former option seems natural to me. Not sure about `ErlangStub` though, maybe just `Erlang` even is it's a stub module.

File: lib/iex/lib/iex/introspection.ex (hunk 12 above)
Comment (eksperimental) [inline lib/iex/lib/iex/introspection.ex]: @bbense sure, 
here is a simple one: https://github.com/elixir-lang/elixir/blob/master/lib/iex/lib/iex/autocomplete.ex#L159-L167

and here a complex one: https://github.com/elixir-lang/elixir/blob/master/lib/mix/lib/mix/tasks/deps.compile.ex#L52-L73

**Comment by bbense** [comment]
Pull request https://github.com/elixir-lang/elixir/pull/3845 is incompatible with this pull request. I'm willing to put in the work to merge the functionality, but I'd like some feedback about whether this functionality is still desired or not.

**Comment by josevalim** [comment]
@bbense we definitely want something like this in but I am not sure the current implementation is the way to go. There is a good amount of indirection (on top of the indirection that already exists) but I can't quite put a finger on it without trying it out myself or by refactoring the existing code (currently in master).

File: lib/iex/lib/iex/dochelp.ex [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+  @type doc_return :: {:found, doc_list} | {:not_found, doc_list} | {:unknown, doc_list}
+
+  @callback documentation(module) :: doc_return
+  @callback documentation(module, function) :: doc_return
+  @callback documentation(module, function, arity) :: doc_return
+
+  @doc """
+  Checks whether `Code.get_docs` should have documentation for the module.
+  """
+  def has_get_docs?(module) do
```
Comment (mgwidmann) [inline lib/iex/lib/iex/dochelp.ex]: This function name is strange since it contains two verbs (has and get)... Maybe consider something like `can_get_docs?`

File: lib/iex/lib/iex/dochelp/elixir.ex [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+    case docs do
+      nil -> {:not_found, [{inspect(module), "No moduledocs found\n"}]}
+      _   -> {:found, [{inspect(module), elem(docs, 1)}]}
+    end
+  end
+
+  defp get_doc(module, function) when is_atom(module) and is_atom(function) do
+    docs = Code.get_docs(module, :docs)
+    case docs do
+      nil -> {:not_found, [{"#{inspect module}.#{function}", "No documentation for #{inspect module}.#{function} found\n"}]}
```
Comment (mgwidmann) [inline lib/iex/lib/iex/dochelp/elixir.ex]: This pattern is repeated a bunch of times in this file. Perhaps it should be a small private function that does this for you in one location.

File: lib/iex/lib/iex/dochelp/erlang_stub.ex [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+  end
+
+  def documentation(module, function,arity) do
+    case erlang?(module) do
+      true -> get_doc(module, function,arity)
+      _    -> {:unknown, [{inspect(module), ""}]}
+    end
+  end
+
+  defp erlang?(module), do: ! has_get_docs?(module)
```
Comment (mgwidmann) [inline lib/iex/lib/iex/dochelp/erlang_stub.ex]: Why not make `has_get_docs?` a callback and implement it here?

File: lib/iex/lib/iex/introspection.ex [hunk 21]
```diff
@@ -13,155 +13,185 @@ defmodule IEx.Introspection do
...
-      else
-        :not_found
-      end
-    else
-      :no_docs
+      _ ->
+        helpers
+        |> Enum.map(fn(mod) -> mod.documentation(module, function) end)
+        |> Enum.filter(fn({status, _doc_list}) -> status != :unknown end)
+        # What about nil result
```
Comment (mgwidmann) [inline lib/iex/lib/iex/introspection.ex]: It wont be nil, you're returning a list... Could be an empty list though...

**Comment by mgwidmann** [comment]
@bbense I would think the logic added in #3845 should just be moved to your helper in `IEx.DocHelp.Elixir`. Since this module is just for Elixir, it should be aware of callbacks and as such be able to handle looking up Elixir documentation correctly.

I'm unclear though why building this pluggable backend is necessary. I can't think of a reason beyond Erlang in which a pluggable backend is useful. Wouldn't it be simpler to just make something like `Code.get_docs` smarter and able to handle looking up Erlang docs (one day)?

**Comment by josevalim** [comment]
@mgwidmann the pluggable backend is necessary because we won't support Erlang soon in get_docs because there is not direct or straight-forward way to retrieve erlang docs (afaik).

**Comment by bbense** [comment]
In order to implement the callback doc redirection, there are two ways to
go.

1. Add the existing logic into doc_help/elixir.ex ( pretty easy )

2. Create a doc_helper that looks for callback documentation and stick it
in the :doc_helper stack.
( also pretty easy, but more code. )

Actually all the stuff with searching [IEx, Kernel, Kernel.SpecialForms]
could also be implemented
as entries in the doc_helper stack.  Much of the complexity in the current
code is around dealing
with this special case. As far as I know there are no functions in common
among the 3 modules so
the :first search syntax would work.

I think with a bit of tweaking you could take all the indirections and turn
it into a single 'search this list of
modules using this function'.

I'm not sure the DocHelper abstraction is exactly correct, but I wanted to
have it duplicate existing
functionality first and then refactor.

**Comment by bbense** [comment]
You can also do things like controlling Dash or opening up a web page or
searching hex. I find the Dash plugin very useful.

I've been down the road of attempting to implement Code.get_docs for Erlang
and it's a giant
can of worms. I've got a mostly working version that works with unix man
pages. See:

github.com/bbense/erlman

But there's no way to create a platform independent solution w/o installing
all the Erlang documentation as part of
elixir.

**Comment by fishcakez** [comment]
> But there's no way to create a platform independent solution w/o installing all the Erlang documentation as part of elixir.

Why didn't we go down this route? It seems a much better approach to me. Could we not build Elixir as an OTP release and parse the xml doc source into markdown and then insert the markdown into the beam files in the release?

**Comment by mgwidmann** [comment]
:+1: this would make Elixir independent from Erlang wouldn't it?

**Comment by fishcakez** [comment]
> :+1: this would make Elixir independent from Erlang wouldn't it?

Well it would mean Elixir would include its own version of Erlang and not use Erlang from path.

**Comment by mgwidmann** [comment]
@fishcakez :+1: I think it'd be a good idea as a default option, with the ability to compile Elixir without Erlang for deployment when as a server (where docs aren't needed).

**Comment by bbense** [comment]
Just to be absolutely clear, even if you have the source XML available you
get at best a buggy incomplete version
of Erlang.get_docs. The xml tags are largely about formatting, not logical
entities. The only way to get documentation
on function X is to apply some heuristics and pattern matching in the
source text; it's actually easier in the nroff output
than in the source XML. The erlang documentation is also not entirely
consistent wrt how those formatting tags are used. I think that's ok as an
external tool you install, but it's not okay for the main distribution.

IMHO, the only solution that would acceptable for distribution in Elixir
would be for h :erlang.foo to open up
the html page for the erlang module, or even just spit back the url.
Anything else is just going to generate an endless
stream of complaints about missing the documentation for function X in
module Y.

**Comment by fishcakez** [comment]
Oh the behaviour is for third party use only? Elixir isn't going to offer any direct support in the stdlib? That's a shame. How feasible would it be to improve the xml upstream for 19.0 so could have native support once we are 19+?

**Comment by bbense** [comment]
Just to double check my assumptions, I looked into how Dash docsets for Erlang are created and it's more or less the same. It just looks for every h3 tag in the html and then classifies it as an anchor type
based on the text. The get_docs problem is actually harder since you need all the text in between anchors and you have to decide if h3 tag is the end of the doc or not. As far as what goes into the stdlib using this functionality, that's up to the Elixir core team.

**Comment by bbense** [comment]
This latest commit duplicates and extends the functionality in PR #3845 
 This allows commands like

h HashSet.size

to print the function description for Set.size. It is somewhat different
from PR #3845 in that it leaves the header as originally parsed from
Code.get_docs. 

One possible addition would be to flag the src of the documentation. 
ie. something like 

                               def size(hash_set)

Returns the number of elements in set.

Examples

┃ iex> Set.size(Enum.into([1, 2, 3], set_impl.new))
┃ 3

This documentation was inferred from Set.size

**Comment by mgwidmann** [comment]
@bbense Initially I put in a preamble indicating the same information, that the docs requested weren't found but that the `Module.whatever` was found. José turned it down, so I assume he'll do the same here but who knows.

File: lib/iex/lib/iex/dochelp/elixir.ex (hunk 19 above)
Comment (bbense) [inline lib/iex/lib/iex/dochelp/elixir.ex]: I wrote one, but I'm not real happy with it. It feels sort of macro like, but that seems like
overkill for just one simple function. I did extract out some helper functions that have cleanup
up this code a lot.

**Comment by bbense** [comment]
I've made a doc check and review pass and squashed all the commits into a single commit.

File: lib/iex/lib/iex/dochelp/erlang_stub.ex [hunk 22]
```diff
@@ -0,0 +1,48 @@
+defmodule IEx.DocHelp.ErlangStub do
+  @moduledoc """
+  Stub module to suggest possible addons for Erlang documentation.
+
+  Currently only returns {:unknown, doc_list }
+  with a suggestive error message.
+  """
+
+  @suggestive "and currently there is no helper installed to provide Erlang documentation"
```
Comment (mgwidmann) [inline lib/iex/lib/iex/dochelp/erlang_stub.ex]: This message makes it appear as if you are missing something that needs to be installed (when I don't believe there is anything that can be installed to remedy this).

Comment (bbense) [inline lib/iex/lib/iex/dochelp/erlang_stub.ex]: But that's the point, you can install something to remedy this, but it will
be an external
platform specific module.

Comment (bbense) [inline lib/iex/lib/iex/dochelp/erlang_stub.ex]: My intent was to only duplicate existing h behavoiur in this PR and then to submit a PR about possible replacements for ErlangStub. See https://github.com/bbense/doc_helpers/blob/master/lib/print_url.ex

for something that might replace ErlangStub.

Comment (mgwidmann) [inline lib/iex/lib/iex/dochelp/erlang_stub.ex]: Perhaps tell the user what command to run or where to look to fix this problem....

**Comment by bbense** [comment]
Is there any point in attempting to resolve the conflicts in this PR? Or should I just close it and move on?

**Comment by josevalim** [comment]
@bbense don't worry. if we are merging it and there are any conflicts, the Elixir team will resolve them before merging, thank you!

**Comment by josevalim** [comment]
Thank you @bbense. After some discussions, we have concluded it would be preferred to allow a solution that would bring its own set of helpers to work with Erlang documentation instead of pigging back on `h`. For example, `e :erlang` instead of `h :erlang`. Sorry for the back and forth on this but it happens on all non-trivial issues.
